documentation/design/proposals/authn_authz_for_1.0.adoc

```diff
@@ -0,0 +1,273 @@
+
+The `AccessControlService` has a selector for namespaces that should use this accesscontrol by default.  This selector is ANDed with the selector of the `MessagingInfra` itself. The `MessagingTenant` may override the the `AccessControlService` by specifying the service's name and namespace.  An annotation of the `AccessControlService` will be used to prevent overriding.
+
+The access control service will use the capabilities of the configured identity providers to determine which SASL
+mechanism(s) to offer during SASL negociation.
+
+Example:
+
+```
+apiVersion: enmasse.io/v1
+kind: AccessControlService
+metadata:
+  name: myaccesscontrol1
+spec:
+  identityProvider:
+    ldap:
+      domain:
+      host:
+      port:
+      tls: # borrow structure from connector (trust, client cert)
+      credentials: # borrow structure from connector
+      searchContext:
+      searchFilter:
+      groupAttributeName: # more fields required to flexibly support LDAP groups.
+    external:
+      domain:
+      host:
+      port:
+      tls: # borrow structure from connector (trust, client cert)
+      credentials: # borrow structure from connector
+    internal:
+      domain:
+    clientAuth:
+      domain:
+    serviceAccount:
+      domain:
+    anonymous:
+      domain:
+  # Allows specifying a selector for namespaces that should use this accesscontrol by default.
+  namespaceSelector:
+    matchLabels:
+      type: messaging
+status:
+  phase: Active
+```
+
+==== MessagingInfra
+
+Configuration provided by `MessagingInfra` will define the Access Control Server deployment.
+
+```
+accessControl:
+    replicas: <integer>
+```
+
+==== MessagingTennant
+
+A `MessagingTennant`can explictly reference a AccessControlService in order to override default configuration provided by the
+Service Admin.  (Should there be a mechanism to prevent overriding?).
+
+```
+accessControlService:
+    name: myaccesscontrol1
+    namespace: custom
+```
+
+==== MessagingUser
+
+`MessagingUser` exists to supports the internal identity source.  It allows a Tenant Admin to define a user/password that
+may connect for messaging.
+
+The stored password will be the Salted Password defined by  https://tools.ietf.org/html/rfc5802. 
+The controller will detect newly created MessagingUser resource and automatically salt plain text passwords.
+
+In this first version, there will be no provision for password policies (constitunency rules, aging etc).
+
+```
+apiVersion: enmasse.io/v1
+kind: MessagingUser
+metadata:
+  name: myuser
+  namespace: mynamespace
+spec:
+  password: {SCRAMSHA256}_saltedpassword_
+```
+
+==== ServiceAccount
+
+The Tenant Admin is able to create an ServiceAccount in the namespace.  Annotating the service account with a special
+annotation enables the user of the service account for messaging authentication.  
```

Comment:
System-wide permissions to read serviceaccounts seems a bit excessive is my first impression. Could we instead accept any token that authenticates a serviceaccount on the cluster, and let the RoleBinding decide what access is granted?

Reply:
In our current system we require the user to nominate `ServiceAccounts` that are eligible for use for messaging authentication using the MessagingUser resources with the special type.  My intent was to keep an explicit step, but simplify it by having the `ServiceAccount` marked with an annotation. That does mean that the AccessControlServer needs cluster wide read permission for ServiceAccounts - firstly to decide if the service account is enabled, and secondly (a nice to have), validate the ServiceAccount references in the MessagingRoleBinding.

If having cluster wide ServiceAccount read is problematic, we can drop the idea of testing the whether the service account is enabled, and drop the validation between ServiceAccount and MessagingRoleBinding (which was probably going to be nothing more than a status message).

Reply:
I think I would prefer to make this configurable at least. Given that Kubernetes does not do any such validation of users referenced in a RoleBinding, I'm not sure if it is the role of the AccessControlServer to do this validation. Maybe it could make sense to delegate this kinf of 'extra' validation to an admission controller instead?

Reply:
I'll drop the annotation from the service account and validation.